src/hotspot/share/runtime/thread.cpp

```diff
@@ -1563,33 +1443,12 @@ void JavaThread::exit(bool destroy_vm, ExitType exit_type) {
       JvmtiExport::post_thread_end(this);
     }
 
-    // We have notified the agents that we are exiting, before we go on,
-    // we must check for a pending external suspend request and honor it
-    // in order to not surprise the thread that made the suspend request.
-    while (true) {
-      {
-        MutexLocker ml(SR_lock(), Mutex::_no_safepoint_check_flag);
-        if (!is_external_suspend()) {
-          set_terminated(_thread_exiting);
-          ThreadService::current_thread_exiting(this, is_daemon(threadObj()));
-          break;
-        }
-        // Implied else:
-        // Things get a little tricky here. We have a pending external
-        // suspend request, but we are holding the SR_lock so we
-        // can't just self-suspend. So we temporarily drop the lock
-        // and then self-suspend.
-      }
-
-      ThreadBlockInVM tbivm(this);
-      java_suspend_self();
+    // The careful dance between thread suspension and exit is handled here.
+    // Since we are in vm and suspension is done with handshakes,
+    // we can just put in the exiting state and it will be corrcetly handled.
```

Comment:
Perhaps: s/Since we are in vm and/Since we are in thread_in_vm state and/
nit typo: s/corrcetly/correctly/

Reply:
Fixed